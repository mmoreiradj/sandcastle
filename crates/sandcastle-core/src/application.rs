use std::path::PathBuf;

use colored::Colorize;
use kube::{Client, Config};
use sandcastle_external_crds::argocd::application::Application;

use crate::{
    Result,
    domain::{
        environment::{
            models::{
                Command, CommentContext, PullRequestContext, ReconcileContext, ReconcileTrigger,
                RepositoryContext, VcsContext, config::SandcastleConfiguration,
            },
            ports::{MockGitOpsPlatformService, MockVCSService},
            services::{ArgoCD, GitOpsPlatform, Vcs},
        },
        repositories::{
            models::GitOpsPlatformType,
            services::{DefaultRepositoryConfigurationService, RepositoryConfigurations},
        },
    },
};

mod http;
mod operator;
pub(crate) mod reconciliation;

const ARGOCD_NAMESPACE: &str = "argocd";

#[derive(Debug, Clone)]
pub struct ApplicationConfig {
    pub argocd_namespace: String,
}

/// State shared beetween the HTTP and Operator
#[derive(Clone)]
pub(crate) struct ApplicationState {
    pub(crate) kube_client: Client,
    pub(crate) namespace: String,
    pub(crate) argocd_namespace: String,
    pub(crate) repository_configuration_service: RepositoryConfigurations,
}

impl ApplicationState {
    pub async fn new(application_config: ApplicationConfig) -> Result<Self, Box<dyn std::error::Error>> {
        let kube_client = Client::try_default().await?;
        let config = Config::infer().await?;
        Ok(Self {
            kube_client,
            namespace: config.default_namespace,
            argocd_namespace: application_config.argocd_namespace,
            repository_configuration_service: DefaultRepositoryConfigurationService::default()
                .into(),
        })
    }

    pub fn operator_context(&self) -> operator::OperatorContext {
        operator::OperatorContext {
            client: self.kube_client.clone(),
            repository_configuration_service: self.repository_configuration_service.clone(),
            namespace: self.namespace.clone(),
        }
    }

    pub fn gitops_platform(&self, gitops_platform_type: &GitOpsPlatformType) -> GitOpsPlatform {
        match gitops_platform_type {
            GitOpsPlatformType::ArgoCD => GitOpsPlatform::ArgoCD(ArgoCD::new(
                self.kube_client.clone(),
                self.argocd_namespace.clone(),
            )),
        }
    }
}

pub async fn start(application_config: ApplicationConfig) -> Result<(), Box<dyn std::error::Error>> {
    let state = ApplicationState::new(application_config).await?;
    let context = state.operator_context();
    tokio::select! {
      _ = http::start(state.clone()) => {
        tracing::info!("HTTP server started");
        Ok(())
      }
      _ = operator::run(context) => {
        tracing::info!("Operator started");
        Ok(())
      }
    }
}

pub async fn test_application(
    file: PathBuf,
    gitops_platform_type: GitOpsPlatformType,
) -> Result<String, Box<dyn std::error::Error>> {
    let content = std::fs::read_to_string(file)?;
    let config = SandcastleConfiguration::from_string(&content)?;
    let context = ReconcileContext {
        config: config.clone(),
        id: "test-123".to_string(),
        vcs: VcsContext {
            repository: RepositoryContext {
                name: "test-repo".to_string(),
                private: false,
                url: "https://github.com/test/repo".to_string(),
            },
            pull_request: PullRequestContext {
                number: 42,
                title: "Test PR".to_string(),
                last_commit_sha: "abc123".to_string(),
            },
            comment: CommentContext {
                body: "test comment".to_string(),
            },
        },
        vcs_service: Vcs::MockVCS(MockVCSService::new()),
        gitops_platform_service: GitOpsPlatform::MockGitOpsPlatform(
            MockGitOpsPlatformService::new(),
        ),
        trigger: ReconcileTrigger::CommentCommand(Command::Deploy),
    };
    let applications = context.template(&config.template)?;
    match gitops_platform_type {
        GitOpsPlatformType::ArgoCD => Ok(applications
            .split("---")
            .map(|application_str| {
                let application_str = application_str.trim();
                let application: Result<Application, serde_yaml::Error> =
                    serde_yaml::from_str(application_str);
                match application {
                    Ok(_) => {
                        format!(
                            r#"{}
{application_str}
"#,
                            "# This application is valid".green(),
                        )
                    }
                    Err(e) => {
                        format!(
                            r#"{}
{application_str}
"#,
                            format!("# This application is invalid: {e:#?}").red(),
                        )
                    }
                }
            })
            .collect::<String>()),
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    #[tokio::test]
    async fn test_test_application() {
        let file = PathBuf::from("tests/fixtures/example_application_2.yaml");
        let applications = test_application(file, GitOpsPlatformType::ArgoCD)
            .await
            .unwrap();
        assert_snapshot!(applications);
    }
}
