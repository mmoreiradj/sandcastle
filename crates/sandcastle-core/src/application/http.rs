use axum_tracing_opentelemetry::middleware::{OtelAxumLayer, OtelInResponseLayer};

pub mod webhook;

/**
 * Start the HTTP server
 */
pub async fn start() {
    let _guard = init_tracing_opentelemetry::TracingConfig::development()
        .init_subscriber()
        .unwrap();
    let router = webhook::router()
        .layer(OtelInResponseLayer)
        .layer(OtelAxumLayer::default());
    let listener = tokio::net::TcpListener::bind("0.0.0.0:8080").await.unwrap();
    axum::serve(listener, router).await.unwrap();
}
