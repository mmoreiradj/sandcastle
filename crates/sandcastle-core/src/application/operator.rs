use std::{sync::Arc, time::Duration};

use futures::StreamExt;
use k8s_openapi::api::core::v1::Secret;
use kube::{
    Api, Client, ResourceExt,
    api::ListParams,
    runtime::{Controller, controller::Action, finalizer},
};
use snafu::ResultExt;
use tracing::instrument;

use crate::{
    Result,
    domain::repositories::{
        models::RepositoryConfiguration, ports::RepositoryConfigurationService,
        services::RepositoryConfigurations,
    },
    error::{FinalizerSnafu, SandcastleError},
    infrastructure::repo_config_service::GithubAppSecretData,
};

#[derive(Clone)]
pub struct OperatorContext {
    pub client: Client,
    pub repository_configuration_service: RepositoryConfigurations,
    pub namespace: String,
}

const SANDCASTLE_FINALIZER: &str = "sandcastle.dev/finalizer";

async fn apply(secret: Arc<Secret>, context: Arc<OperatorContext>) -> Result<Action> {
    let github_app_secret_data = GithubAppSecretData::from_secret(secret)?;
    let repository_config = RepositoryConfiguration::from(github_app_secret_data);
    context
        .repository_configuration_service
        .upsert_repository_configuration(repository_config)
        .await?;
    Ok(Action::requeue(Duration::from_secs(5 * 60)))
}

async fn cleanup(secret: Arc<Secret>, context: Arc<OperatorContext>) -> Result<Action> {
    context
        .repository_configuration_service
        .delete_repository_configuration(&secret.name_any())
        .await?;
    Ok(Action::requeue(Duration::from_secs(5 * 60)))
}

#[instrument(skip(secret, context), fields(_secret_name))]
async fn reconcile(secret: Arc<Secret>, context: Arc<OperatorContext>) -> Result<Action> {
    let _secret_name = &secret.name_any();
    tracing::info!("Reconciling secret");
    let secrets = Api::<Secret>::namespaced(context.client.clone(), &context.namespace);

    finalizer(&secrets, SANDCASTLE_FINALIZER, secret, |event| async move {
        match event {
            finalizer::Event::Apply(secret) => apply(secret, context.clone()).await,
            finalizer::Event::Cleanup(secret) => cleanup(secret, context.clone()).await,
        }
    })
    .await
    .context(FinalizerSnafu)
}

#[instrument(skip(secret, error, _context), fields(_secret_name))]
fn error_policy(
    secret: Arc<Secret>,
    error: &SandcastleError,
    _context: Arc<OperatorContext>,
) -> Action {
    let _secret_name = secret.name_any();
    tracing::warn!("Failed reconciling secret {:?}", error);
    Action::requeue(Duration::from_secs(5 * 60))
}

pub async fn run(client: Client, context: OperatorContext) {
    let secrets = Api::<Secret>::namespaced(client.clone(), &context.namespace);

    if let Err(e) = secrets.list(&ListParams::default().limit(1)).await {
        tracing::error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }
    let watcher_config =
        kube::runtime::watcher::Config::default().labels("sandcastle.dev/secret-type=repository");
    Controller::new(secrets, watcher_config.clone())
        .shutdown_on_signal()
        .run(reconcile, error_policy, Arc::new(context.clone()))
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
