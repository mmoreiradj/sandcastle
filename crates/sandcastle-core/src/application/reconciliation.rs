use tracing::{info, instrument};

use crate::{
    Result,
    domain::environment::models::{
        Command, CreateOrUpdateArgocdApplicationAction, DeleteArgocdApplicationAction,
        ReconcileActions, ReconcileContext, ReconcileTrigger,
    },
    error::SandcastleError,
};

pub struct ReconciliationService;

impl ReconciliationService {
    #[instrument(skip(context))]
    pub async fn reconcile(context: ReconcileContext) -> Result<(), SandcastleError> {
        info!("Processing reconciliation trigger: {:?}", context.trigger);

        let action = Self::determine_action(&context)?;

        info!("Executing action: {:?}", action);

        action.reconcile(context).await?;

        info!("Reconciliation completed successfully");

        Ok(())
    }

    fn determine_action(context: &ReconcileContext) -> Result<ReconcileActions, SandcastleError> {
        let templated_application = context.template(&context.config.template)?;

        match &context.trigger {
            ReconcileTrigger::CommentCommand(Command::Deploy) => {
                Ok(ReconcileActions::CreateOrUpdateArgocdApplication(
                    CreateOrUpdateArgocdApplicationAction {
                        application: templated_application,
                    },
                ))
            }
            ReconcileTrigger::CommentCommand(Command::Destroy) => Ok(
                ReconcileActions::DeleteArgocdApplication(DeleteArgocdApplicationAction {
                    application: templated_application,
                }),
            ),
            ReconcileTrigger::PushEvent => Ok(ReconcileActions::CreateOrUpdateArgocdApplication(
                CreateOrUpdateArgocdApplicationAction {
                    application: templated_application,
                },
            )),
            ReconcileTrigger::PullRequestClosed => Ok(ReconcileActions::DeleteArgocdApplication(
                DeleteArgocdApplicationAction {
                    application: templated_application,
                },
            )),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::environment::{
        models::{
            CommentContext, PullRequestContext, RepositoryContext, config::SandcastleConfiguration,
            VcsContext,
        },
        ports::MockVCSService,
        services::{ArgoCD, GitOpsPlatform, VCS},
    };

    fn test_context(trigger: ReconcileTrigger) -> ReconcileContext {
        let config = SandcastleConfiguration::from_string(include_str!(
            "../../tests/fixtures/example_application_1.yaml"
        ))
        .unwrap();

        ReconcileContext {
            id: "test-123".to_string(),
            vcs: VcsContext {
                repository: RepositoryContext {
                    name: "test-repo".to_string(),
                    private: false,
                    url: "https://github.com/test/repo".to_string(),
                },
                pull_request: PullRequestContext {
                    number: 42,
                    title: "Test PR".to_string(),
                    last_commit_sha: "abc123".to_string(),
                },
                comment: CommentContext {
                    body: String::new(),
                },
            },
            vcs_service: VCS::MockVCS(MockVCSService::new()),
            gitops_platform_service: GitOpsPlatform::ArgoCD(ArgoCD),
            config,
            trigger,
        }
    }

    #[tokio::test]
    async fn test_reconcile_deploy_command() {
        let context = test_context(ReconcileTrigger::CommentCommand(Command::Deploy));
        let action = ReconciliationService::determine_action(&context).unwrap();

        match action {
            ReconcileActions::CreateOrUpdateArgocdApplication(action) => {
                assert!(action.application.contains("kind: Application"));
                assert!(action.application.contains("frontend-test-repo"));
            }
            _ => panic!("Expected CreateOrUpdateArgocdApplication action"),
        }
    }

    #[tokio::test]
    async fn test_reconcile_destroy_command() {
        let context = test_context(ReconcileTrigger::CommentCommand(Command::Destroy));
        let action = ReconciliationService::determine_action(&context).unwrap();

        match action {
            ReconcileActions::DeleteArgocdApplication(action) => {
                assert!(action.application.contains("kind: Application"));
            }
            _ => panic!("Expected DeleteArgocdApplication action"),
        }
    }

    #[tokio::test]
    async fn test_reconcile_push_event() {
        let context = test_context(ReconcileTrigger::PushEvent);
        let action = ReconciliationService::determine_action(&context).unwrap();

        match action {
            ReconcileActions::CreateOrUpdateArgocdApplication(action) => {
                assert!(action.application.contains("kind: Application"));
            }
            _ => panic!("Expected CreateOrUpdateArgocdApplication action"),
        }
    }

    #[tokio::test]
    async fn test_reconcile_pr_closed() {
        let context = test_context(ReconcileTrigger::PullRequestClosed);
        let action = ReconciliationService::determine_action(&context).unwrap();

        match action {
            ReconcileActions::DeleteArgocdApplication(action) => {
                assert!(action.application.contains("kind: Application"));
            }
            _ => panic!("Expected DeleteArgocdApplication action"),
        }
    }
}
