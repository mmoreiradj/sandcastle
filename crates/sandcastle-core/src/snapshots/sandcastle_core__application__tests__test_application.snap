---
source: crates/sandcastle-core/src/application.rs
expression: applications
---
[32m# This application is valid[0m
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: frontend-test-repo-42
  namespace: argocd
spec:
  project: default
  source:
    path: "charts/frontend" 
    repoURL: "https://github.com/test/repo"
    targetRevision: "abc123"
    helm:
      valuesObject:
        ingress:
          enabled: true
          className: "traefik"
          hosts:
            - host: "frontend-test-repo-42.sandcastle.dev"
              paths:
                - path: /
                  pathType: ImplementationSpecific
        image:
          tag: "abc123"
  destination:
    server: "https://kubernetes.default.svc"
    namespace: "test-repo-42"
